pkg/image/prune/prune.go

```diff
@@ -685,57 +696,68 @@ func pruneStreams(g graph.Graph, imageNodes []*imagegraph.ImageNode, streamPrune
-				if err != nil {
-					return err
-				}
-				updatedTags := sets.NewString()
-
-				glog.V(4).Infof("Checking if ImageStream %s has references to image %s in status.tags", getName(stream), imageNode.Image.Name)
+			streamName := getName(streamNode.ImageStream)
+			prunableStreams[streamName] = streamNode
+		}
+	}
 
-				for tag, history := range stream.Status.Tags {
-					glog.V(4).Infof("Checking tag %q", tag)
+	for streamName, streamNode := range prunableStreams {
+		err := retry.RetryOnConflict(retry.DefaultRetry, func() error {
+			stream, err := streamPruner.GetImageStream(streamNode.ImageStream)
+			if err != nil {
+				// During prune someone removed imagestream
+				if kerrors.IsNotFound(err) {
+					return nil
+				}
+				return err
+			}
 
-					newHistory := imageapi.TagEventList{}
+			updatedTags := sets.NewString()
+			deletedTags := sets.NewString()
 
-					for i, tagEvent := range history.Items {
-						glog.V(4).Infof("Checking tag event %d with image %q", i, tagEvent.Image)
+			for tag, history := range stream.Status.Tags {
+				newHistory := imageapi.TagEventList{}
+				for i, tagEvent := range history.Items {
+					glog.V(4).Infof("Checking tag event %d with image %q", i, tagEvent.Image)
 
-						if tagEvent.Image != imageNode.Image.Name {
-							glog.V(4).Infof("Tag event doesn't match deleted image - keeping")
-							newHistory.Items = append(newHistory.Items, tagEvent)
-						} else {
-							glog.V(4).Infof("Tag event matches deleted image - removing reference")
-							updatedTags.Insert(tag)
-						}
-					}
-					if len(newHistory.Items) == 0 {
-						glog.V(4).Infof("Removing tag %q from status.tags of ImageStream %s", tag, getName(stream))
-						delete(stream.Status.Tags, tag)
+					if _, ok := imageNodes[tagEvent.Image]; !ok {
+						glog.V(4).Infof("Image stream tag %s:%s revision %d - keeping because image %q is not deleted", streamName, tag, i, tagEvent.Image)
+						newHistory.Items = append(newHistory.Items, tagEvent)
 					} else {
-						stream.Status.Tags[tag] = newHistory
+						glog.V(4).Infof("Image stream tag %s:%s revision %d - removing because image %q matches deleted image", streamName, tag, i, tagEvent.Image)
+						updatedTags.Insert(tag)
 					}
 				}
 
-				updatedStream, err := streamPruner.UpdateImageStream(stream, imageNode.Image, updatedTags.List())
-				if err == nil {
-					streamNode.ImageStream = updatedStream
+				if len(newHistory.Items) == 0 {
+					glog.V(4).Infof("Image stream tag %s:%s - removing empty tag", streamName, tag)
+					delete(stream.Status.Tags, tag)
+					deletedTags.Insert(tag)
+				} else {
+					stream.Status.Tags[tag] = newHistory
 				}
-				return err
-			}); err != nil {
-				errs = append(errs, fmt.Errorf("error removing image %s from stream %s: %v",
-					imageNode.Image.Name, getName(streamNode.ImageStream), err))
-				continue
 			}
+
+			updatedStream, err := streamPruner.UpdateImageStream(stream)
```

Comment:
we should know if we actually changed the stream or not, can't we be smarter here?

Reply:
Yep. Thanks!

Reply:
Though, on the previous step we've selected all image stream that should be modified. If the stream is not changed here, it means someone else concurrently removes objects from the image stream.